crates/ingress-kafka/src/dispatcher.rs

```diff
@@ -0,0 +1,193 @@
+            invocation_id,
+            invocation_target,
+            restate_types::invocation::Source::Ingress(PartitionProcessorRpcRequestId::new()),
+        );
+        service_invocation.with_related_span(related_span);
+        service_invocation.argument = argument;
+        service_invocation.headers = headers;
+
+        Ok(KafkaIngressEvent {
+            service_invocation,
+            deduplication_id,
+            deduplication_index,
+            proxying_partition_key,
+        })
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum IngressDispatchError {
+    #[error("bifrost error: {0}")]
+    WalProtocol(#[from] restate_wal_protocol::Error),
+    #[error("partition routing error: {0}")]
+    PartitionRoutingError(#[from] PartitionTableError),
+}
+
+/// Dispatches a request from kafka ingress to bifrost
+pub trait DispatchKafkaEvent {
+    fn dispatch_kafka_event(
+        &self,
+        event: KafkaIngressEvent,
+    ) -> impl std::future::Future<Output = Result<(), IngressDispatchError>> + Send;
+}
+
+#[derive(Clone)]
+pub(crate) struct KafkaIngressDispatcher {
+    bifrost: Bifrost,
+}
+
+impl KafkaIngressDispatcher {
+    pub(crate) fn new(bifrost: Bifrost) -> Self {
+        Self { bifrost }
+    }
+}
+
+impl DispatchKafkaEvent for KafkaIngressDispatcher {
+    async fn dispatch_kafka_event(
+        &self,
+        ingress_request: KafkaIngressEvent,
+    ) -> Result<(), IngressDispatchError> {
+        let KafkaIngressEvent {
+            service_invocation: inner,
+            deduplication_id,
+            deduplication_index,
+            proxying_partition_key,
+        } = ingress_request;
+
+        let partition_key = proxying_partition_key.unwrap_or_else(|| inner.partition_key());
+
+        let envelope = wrap_service_invocation_in_envelope(
+            partition_key,
+            inner,
+            metadata().my_node_id(),
+            deduplication_id.to_string(),
+            deduplication_index,
+        );
+        let (log_id, lsn) = append_envelope_to_bifrost(&self.bifrost, Arc::new(envelope)).await?;
+
+        debug!(
+            log_id = %log_id,
+            lsn = %lsn,
+            "Ingress request written to bifrost"
+        );
+        Ok(())
+    }
+}
+
+fn wrap_service_invocation_in_envelope(
+    partition_key: PartitionKey,
+    service_invocation: ServiceInvocation,
+    from_node_id: GenerationalNodeId,
+    deduplication_source: String,
+    deduplication_index: MessageIndex,
+) -> Envelope {
+    let header = Header {
+        source: Source::Ingress {
+            node_id: from_node_id,
+            nodes_config_version: metadata().nodes_config_version(),
+        },
+        dest: Destination::Processor {
+            partition_key,
+            dedup: Some(DedupInformation::ingress(
+                deduplication_source,
+                deduplication_index,
+            )),
+        },
+    };
+
+    Envelope::new(header, Command::Invoke(service_invocation))
```

Comment:
Does this mean that we can remove the `Command::ProxyThrough` completely? Right now it seems that it is no longer used.

Originally, we introduced the `ProxyThrough` command so that a Kafka ingress won't have to talk to every partition (to avoid growing the deduplication map too much: O(m * n) with m being the number of partitions and n being the number of topic partitions). Now, this seems to be changed. I am not sure whether this is real problem, though.

Reply:
No sorry the PR was wrong, i removed the wrong thing :D